### ISSUE TITLE
Overview of project

### ISSUE THREAD

**Issue by jflasher** [openaq/voltaire#1]
I've been thinking about this a bit more and I still like my initial idea in https://github.com/openaq/openaq-api/issues/366 (surprise!). [S3 Select](https://aws.amazon.com/blogs/aws/s3-glacier-select/) (in preview) should really help with this as well as it'll mean a lot of the data pruning will be done by S3 and mean less data that needs to be moved through the Lambda function. Some more thoughts on what this could look like below.

- Create standalone Lambda function (probably using http://apex.run/) that handles the query 
    1. Make sure this is in the same region as the data (us-east-1)
    1. Eventually we'll need both a staging and production tags (can be handled with Apex)
    1. The Lambda function should take query parameters (probably similar to how we do via openaq-api?), loop over every file in `openaq-fetches/realtime/`, filter for matching measurements, save them to a file and write that file to an output S3 bucket.
    1. There is a 3GB memory limit and a 5 minute time limit, if the average result sets do not fit within that, we may need to use something like https://aws.amazon.com/batch/, but I'd like to try and start with Lambda.
    1. We'll need to create a new S3 bucket where these output files can go, we should set a lifecycle policy so the files are deleted after 1 day so we don't just end up storing a lot of old files there.
    1. Because we're depending on the speed of running the Lambda function in the same region as the data, testing this locally won't really make sense unless it's only looking at a subset of the data. The timing we see when running locally will be much slower than seen when running on AWS.
    1. We can still look at this solution even if S3 Select isn't available right away, it just means the full files would need to be pulled onto the Lambda function and the filtering done via the Lambda function instead of S3.
- Once the above Lambda function is in place, we'll probably want to add a new endpoint to `openaq-api` that's just a wrapper around the functionality. It's job is to collect the query parameters, start the lambda function and immediately return the final S3 URL of the output file.
- I want `openaq-api` to immediately return the final S3 url for the results file so that the user can then keep checking for that file to be present. To do this, either 1) `openaq-api` needs to tell the Lambda function what the final URL is or 2) the Lambda function needs to somehow return data immediately while still running behind the scenes to do the work of filtering the data (I don't think this is possible). If we use a UUID generator to create the final URL, I don't think 1) should be a problem.

cc/ @dolugen

**Comment by dolugen** [comment]
Let's get it started!

My first goal is to get a basic lambda function running, which accepts a bucket, key and an SQL expression, and saves the output to a different bucket. The S3 Select sample code is a good starting point.

Gotcha: Lambda expects `main.handle` with Python3.6.
Maybe bug: `select_object_content`'s CSV output wasn't working for me.
Docs: http://s3select.s3-website-us-east-1.amazonaws.com

**Comment by dolugen** [comment]
I have a function that runs well on _one_ object. How can I make it run on hundreds and thousands of objects? Can I do it with Lambda alone?

Here's what I've tried so far with my Lambda function:
1. Tried using `multiprocessing.Pipe`. I didn't get it working.
2. Used `ThreadPoolExecutor` from `concurrent.futures`. It works on 500 objects, but not 1000 (memory limit exceeded).

There's an idea that still uses Lambda, but the above function would be one of many. It comes from [this post](https://www.awsadvent.com/2016/12/04/exploring-concurrency-in-python-aws/) and a talk with Joe about a "fanout" method. The gist of it is this: A query sends a number of SNS messages that trigger the lambda functions, which each generates an output from one or a few objects, and somehow they're merged into one file at the end.

**Comment by jflasher** [comment]
@dolugen Asked some folks over at Dev Seed for ideas and they pointed to http://eventlet.net/ as something that might help with handling the concurrency. I imagine we'll still need to do some about of fanout but it'd be good to pack as many requests in one Lambda function as possible (1 day of objects, 1 week? etc).

cc/ @drewbo @scisco @wronk

**Comment by dolugen** [comment]
Thanks @jflasher, I'll try that out.

**Comment by jflasher** [comment]
@dolugen I think we may really be able to simplify this whole thing by using Athena. Then we can skip hitting all the objects and let the service handle it. I think the overall idea is still the same though.

**Comment by dolugen** [comment]
Right! It's viable now that you've made optimizations. I'll look into that.

**Comment by jflasher** [comment]
Awesome, thanks!

**Comment by rickpeltier** [comment]
Using Athena to query data into an S3 bucket (as .csv) is a fairly straightforward task, once I went through the process a few times.  We did learn that Athena times out at 30 minutes for a query, and my interest was to export all of  the openaq data to date.  As a result, an unfiltered query (all data, all sites) that was anything more than ~1 year (e.g. all data from 2017 and 2018) just barely made it under the timeout clock.  But, a query for all data from 2000-2014 (with much more limited openaq data in this timeframe, of course) took 10s of seconds.  

As of now, it queries the gzipped version of the data in S3, which is around 3GB.  Probably ran the query 10 times until I figured out the timeout issue and broke the queries into separate ones.  Should note that any query you run will be based on this ~3GB+ archive, and that's how AWS will assign charges; and it still cost less than $0.01 in total.  The biggest downside is that the table created from the query is uncompressed in my S3 bucket, and can not be compressed unless you move it to an EC2 instance.  For those interested, the full dataset in an uncompressed format is about 50GB (as of now).  

I think the real benefit here is to be able to run queries in the background automatically, and use this output to keep databases current without user input.

The output format is a little wonky and will need to be cleaned up, but all doable.

**Comment by dolugen** [comment]
@rickpeltier Thank you for the details of your account, it's definitely interesting! Especially the part that Athena had time limits of 30 mins, I didn't know that.

I'm not sure at this point, whether to break down the queries ourselves, or let whoever's using voltaire do it. It'll be used in the API, so we could constrain by a maximum of 1 year date range.

Also, will be looking at compressing the Athena output, so we don't need to keep around large output files, and the user can use less bandwidth. This could be another lambda like `fetches_optimizer`.

**Comment by RocketD0g** [comment]
cc: @abarciauskas-bgse, @AreteY
